### DISCUSSION TITLE
[help] error fetching AWS S3 file targets with recent targets versions

### DISCUSSION THREAD

**Discussion by jds485** [ropensci/targets#1102] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I recently updated from targets 0.11.0 to the latest version, 1.2.0. With that update, I am unable to `tar_load()` any file target that is stored on AWS S3. This happens for targets that were generated with 0.11.0 or new targets made with 1.2.0. I also tried with targets version 0.14.3 and had the same issue. I am not sure if this is a `targets` issue, but I saw a similar error message in #410 and wanted to check in.

A simple repex is to create a file target with targets version 0.14.3 or 1.2.0 and try to `tar_load()` the file target:
```
  tar_target(
    test_aws_file,
    {
      write.csv(data.frame(a=1), 'test.csv')
      'test.csv'
    },
    format = "file",
    repository = "aws"
  ),
  tar_target(
    test_aws_file_depend,
    read.csv(test_aws_file),
    repository = "aws"
  )
tar_make(test_aws_file) #works and the file is showing in the S3 bucket
tar_load(test_aws_file) #warning message and file doesn’t appear in my directory
tar_make(test_aws_file_depend) #fails with error message below because the file does not load from S3
```

For me, that results in this warning for the ` tar_load(test_aws_file)`
```
Warning message:
In file.rename(from = scratch, to = stage) :
  cannot rename file '/tmp/RtmpKQBTDI/scratch/targets_aws_file_1446a6e99e5' to 'test.csv', reason 'Invalid cross-device link'
```

And this error message for ` tar_make(test_aws_file_depend)`
```
"cannot open file test.csv No such file or directory"
Error:
! Error running targets::tar_make()
  Error messages: targets::tar_meta(fields = error, complete_only = TRUE)
  Debugging guide: https://books.ropensci.org/targets/debugging.html
  How to ask for help: https://books.ropensci.org/targets/help.html
  Last error: cannot open the connection
```

This is being run in R version 4.1.3.

**Comment by wlandau** [comment]
The "Invalid cross-device link" error is often related to a file system operation and can happen when you're trying to move a file across different filesystems or partitions. This is common in Unix-like systems where the rename() system call is used to move a file, which requires the source and destination to be on the same mounted filesystem.  For AWS file targets, intermediate files are staged in `tempdir()` and then moved from there with `file.rename()`, which uses Unix `rename()`. So I might have to patch `targets` to work around that somehow, either avoiding `tempdir()` or doing a safer file copy.

**Reply by jds485** [reply]
Thanks! I am working in Linux (Ubuntu 22.04)

**Reply by wlandau** [reply]
Does `tempdir()` live on a different physical drive than your usual workspace?

**Reply by jds485** [reply]
I'm not sure. But when I type `tempdir()` in R and try to `cd` to the folder in the command line, it does not work. So, maybe it is a different drive? If it helps, I've tried this within a Docker container on WSL2, and also in a Singularity container on a Linux HPC.

**Reply by wlandau** [reply]
As long as the R session that created `tempdir()` is still open, that directory should still exist and `cd` should work. Seems like like a system issue. But I will see if I can have `targets` download to the `scratch` folder in the data store instead of a temp file.

**Reply by wlandau** [reply]
@jds485, this should be fixed as of 8c4f85699263d9d1f096e8a01e6484c003970378. I used `fs::file_move()` which is robust when files are on different drives, and I changed the scratch/staging area for uploads/downloads from `tempdir()` to `tools::R_user_dir("targets", "cache")`.

**Reply by jds485** [reply]
Thanks! targets version 1.2.1 worked for me in Docker on WSL2. In Singularity on HPC I still am unable to tar_load the file target (same warning as above with a different initial file path: 

```
> tar_load(test_aws_file)
Warning message:
In file.rename(from = from, to = to) :
  cannot rename file '/home/<username>/.cache/R/targets/scratch/test_aws_file' to 'test.csv', reason 'Invalid cross-device link'
```

Surprisingly, the dependency object target builds successfully. I'm not sure how that's possible. I checked to make sure that the `test.csv` is not already in the working directory. I see that you changed to `file_move`, but the warning message here says `file.rename` - is it possible that there are other locations in the repo with `file.rename` that are causing this warning and inability to load the file?

**Reply by wlandau** [reply]
Just realized the copy fallback wasn't working since file.rename() throws a warning instead of an error when it fails. That part is fixed in c999968e182d232f517c0a9eb5bd24c75b197cdd, so please give that a try. 

Still surprised you are having trouble moving files. `targets` now stages files in the user's home directory instead of inside `tempdir()`, so it's odd that it doesn't work for you.

**Reply by wlandau** [reply]
In general, do you have trouble copying files from ~/.cache/ to wherever you are writing `test.csv`?

**Reply by jds485** [reply]
Thanks! Version `1.2.2.9000` works for me. 

> In general, do you have trouble copying files from ~/.cache/ to wherever you are writing test.csv?

I just tried. Yes with `file.rename`, no with `fs::file_move`. The `file.rename` warning message is the same as above.
